SetupDataPkg/Tools/KnobService.py

```diff
@@ -0,0 +1,571 @@
+            out.write("{} config_get_{}();\n".format(
+                get_type_string(knob.format.c_type, efi_type),
+                knob.name))
+            out.write("\n")
+
+            out.write("#ifdef CONFIG_SET_VARIABLES\n")
+            out.write("// Set the current value of the {} knob\n".format(knob.name))
+            out.write("{} config_set_{}({} value);\n".format(
+                get_type_string('bool', efi_type),
+                knob.name,
+                get_type_string(knob.format.c_type, efi_type),
+            ))
+            out.write("#endif // CONFIG_SET_VARIABLES\n")
+            out.write("\n")
+            pass
+
+        out.write("\n")
+        out.write("typedef enum {\n")
+        for knob in schema.knobs:
+            out.write("    KNOB_{},\n".format(knob.name))
+        out.write("    KNOB_MAX\n")
+        out.write("} knob_t;\n")
+        out.write("\n")
+        if efi_type:
+            out.write("typedef EFI_GUID config_guid_t;\n")
+        else:
+            out.write("typedef struct {\n")
+            out.write("    unsigned long  Data1;\n")
+            out.write("    unsigned short Data2;\n")
+            out.write("    unsigned short Data3;\n")
+            out.write("    unsigned char  Data4[8];\n")
+            out.write("} config_guid_t;\n")
+        out.write("\n")
+        out.write("typedef struct {\n")
+        out.write("    int get_count;\n")
+        out.write("    int set_count;\n")
+        out.write("} knob_statistics_t;\n")
+        out.write("\n")
+        out.write(f"typedef {get_type_string('bool', efi_type)}(knob_validation_fn)(const void*);")
+        out.write("\n")
+        out.write("typedef struct {\n")
+        out.write("    knob_t knob;\n")
+        out.write("    const void* default_value_address;\n")
+        out.write("    void* cache_value_address;\n")
+        out.write("    {} value_size;\n".format(get_type_string('size_t', efi_type)))
+        out.write("    const char* name;\n")
+        out.write("    {} name_size;\n".format(get_type_string('size_t', efi_type)))
+        out.write("    config_guid_t vendor_namespace;\n")
+        out.write("    int attributes;\n")
+        out.write("    knob_statistics_t statistics;\n")
+        out.write("    knob_validation_fn* validator;\n")
+        out.write("} knob_data_t;\n")
+        out.write("\n")
+        out.write("typedef struct {\n")
+        out.write("    knob_t knob;\n")
+        out.write("    void* value;\n")
+        out.write("} knob_override_t;\n")
+        out.write("\n")
+        out.write("typedef struct {\n")
+        out.write("    knob_override_t* overrides;\n")
+        out.write("    {} override_count;\n".format(get_type_string('size_t', efi_type)))
+        out.write("} profile_t;\n")
+        out.write("\n")
+        out.write("#pragma pack(pop)")
+        out.write("\n")
+
+
+def format_guid(guid):
+    u = uuid.UUID(guid)

```

Comment:
Should this use little endian (bytes_le)? Just need to be consistent throughout the system: https://uefi.org/specs/UEFI/2.10/Apx_A_GUID_and_Time_Formats.html?highlight=guid.

Reply:
I will test this.

Reply:
guid comes in here as a string of the GUID as written in the XML, whereas bytes_le takes in a byte string. I think the code below this is doing the conversion correctly, but let me analyze it.